### PR TITLE
Add client-side password generator to create user / change password page

### PR DESCRIPTION
**Pull request by IniZio**
ref DEV-1713

dependencies:
- [x] #4535

to be confirmed:
- ~~Whether to use fluentui's suffix slot for generate button and reveal button~~

### Screens

Name|Screenshot
-|-
Change Password|image
Create User|image

### CODE REVIEW

**Comment by louischan-oursky** [comment]
Please rebase.

**Comment by IniZio** [comment]
Rebased 🙏

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: portal/src/util/passwordGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+const MaxTrials = 10;
+const DefaultMinLength = 8;
+const GuessableEnabledMinLength = 32;
+
+export interface RandSource {
+  pick(list: string): number;
+  shuffle(list: string): string;
+}
+
+class CryptoRandSource implements RandSource {
```
[the remark this answers is not in the document]
Comment (IniZio) [reply portal/src/util/passwordGenerator.ts]: Updated :adore2x:

File: portal/src/util/passwordGenerator.ts [hunk 2]
```diff
@@ -19,96 +19,93 @@ export interface RandSource {
...
+    const array = list.split("");
+    for (let i = array.length - 1; i > 0; i--) {
+      const randomArray = new Uint8Array(1);
+      window.crypto.getRandomValues(randomArray);
+      const j = randomArray[0] % (i + 1);
+      [array[i], array[j]] = [array[j], array[i]];
+    }
+    return array.join("");
+  },
+};
```
Comment (louischan-oursky) [inline portal/src/util/passwordGenerator.ts:38]:
```
interface Source {
  // Return [0, n). Assume n <= 256
  intn(n: number): number;
}
```

This is Intn in JavaScript with Math.random https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random#getting_a_random_integer_between_two_values

```
const MAX_SAFE_INTEGER = Math.pow(2, 32) - 1;

if (n > MAX_SAFE_INTEGER) {
  throw new Error();
}
const numberBytesNeeded = Math.ceil(Math.log2(n) / 8);
forever {
  const array = new Uint8Array(numberBytesNeeded);
  window.crypto.getRandomValues(array);
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView/getUint32
  // Handle modular bias
}
```

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: portal/src/util/passwordGenerator.ts (hunk 2 above)
Comment (IniZio) [reply portal/src/util/passwordGenerator.ts:42]: Updated.

File: portal/src/util/passwordGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+  if (set.size === 0) {
+    set.add(CharListAlphabet);
+    set.add(CharListDigit);
+  }
+
+  // Remove overlapping character sets.
+  if (set.has(CharListAlphabet)) {
+    set.delete(CharListLowercase);
+    set.delete(CharListUppercase);
+  }
```
Comment (louischan-oursky) [inline portal/src/util/passwordGenerator.ts:150]: This does not match the server implementation.

File: portal/src/util/passwordGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+  // Build the final character list.
+  let charList = "";
+  set.forEach((cs) => {
```
Comment (louischan-oursky) [inline portal/src/util/passwordGenerator.ts:154]: This does not match the server implementation. The server implementation uses allCharacterSets list to ensure a predictable order.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: portal/src/util/passwordGenerator.ts (hunk 4 above)
Comment (IniZio) [reply portal/src/util/passwordGenerator.ts]: Rebased and updated in [c4eabcd](https://github.com/authgear/authgear-server/pull/4537/commits/c4eabcdcafbc2f50cf4410c3e900a62300bb66ff)

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
1. When the user has only a phone number
SCR-20240812-pqrm

2. We have an unexpected warning
SCR-20240812-pqxb

**Comment by louischan-oursky** [comment]
@IniZio Prioritize this issue first.

**Comment by IniZio** [comment]
@louischan-oursky Updated 🙏

File: portal/src/graphql/adminapi/ChangePasswordScreen.tsx [hunk 5]
```diff
@@ -131,19 +131,21 @@ const ChangePasswordContent: React.VFC<ResetPasswordContentProps> = function (
...
-            options={passwordCreateionTypeOptions}
-            onChange={onChangePasswordCreationType}
-          />
-        </div>
+        {user?.standardAttributes?.email != null ? (
+          <div>
+            <TextField
+              label={renderToString("ChangePasswordScreen.email")}
+              type="email"
+              value={user.standardAttributes.email}
```
Comment (louischan-oursky) [inline portal/src/graphql/adminapi/ChangePasswordScreen.tsx:139]: Condition to show this:

1. When the user has a email Login ID. Having a email standard attribute is not enough.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by louischan-oursky** [comment]
@IniZio I added one more commit to fix

1. Disable the send password checkbox when there is no email. Previously the checkbox is enabled, and if it is enabled, we will encounter the No Email Authenticator error.
2. Move the useQuery to the screen and block until it is loaded.

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**